# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsRefreshDriver.cpp
# Commit: 95679f6ce254
# Full Hash: 95679f6ce2544b21ec9cfd0d19961058d60155d6
# Author: Frederic Wang <fred.wang@free.fr>
# Date: 2023-11-27 17:00:06
# Regressor Bug: 1807253
# File Overlap Count: 1
# Description:
#   Bug 1807253 - unreliable timing for content-visibility:auto, r=emilio
#   
#   Rewrite implementation of content-visibility: auto as defined in
#   https://github.com/w3c/csswg-drafts/issues/8542
#   
# ==============================================================================

diff -r 9ea04a2cc829 -r 95679f6ce254 layout/base/nsRefreshDriver.cpp
--- a/layout/base/nsRefreshDriver.cpp	Mon Nov 27 08:00:41 2023 +0000
+++ b/layout/base/nsRefreshDriver.cpp	Mon Nov 27 08:26:46 2023 +0000
@@ -2242,11 +2242,8 @@
   mNeedToUpdateContentRelevancy = false;
 }
 
-void nsRefreshDriver::NotifyResizeObservers() {
-  AUTO_PROFILER_LABEL_RELEVANT_FOR_JS("Notify ResizeObserver", LAYOUT);
-  if (!mNeedToUpdateResizeObservers) {
-    return;
-  }
+void nsRefreshDriver::DetermineProximityToViewportAndNotifyResizeObservers() {
+  AUTO_PROFILER_LABEL_RELEVANT_FOR_JS("Update the rendering: step 14", LAYOUT);
   // NotifyResizeObservers might re-schedule us for next tick.
   mNeedToUpdateResizeObservers = false;
 
@@ -2255,17 +2252,19 @@
   }
 
   AutoTArray<RefPtr<Document>, 32> documents;
-  if (mPresContext->Document()->HasResizeObservers()) {
+  if (mPresContext->Document()->HasResizeObservers() ||
+      mPresContext->Document()->HasContentVisibilityAutoElements()) {
     documents.AppendElement(mPresContext->Document());
   }
 
   mPresContext->Document()->CollectDescendantDocuments(
       documents, [](const Document* document) -> bool {
-        return document->HasResizeObservers();
+        return document->HasResizeObservers() ||
+               document->HasContentVisibilityAutoElements();
       });
 
   for (const RefPtr<Document>& doc : documents) {
-    MOZ_KnownLive(doc)->NotifyResizeObservers();
+    MOZ_KnownLive(doc)->DetermineProximityToViewportAndNotifyResizeObservers();
   }
 }
 
@@ -2748,15 +2747,6 @@
     pm->UpdatePopupPositions(this);
   }
 
-  // Notify resize observers if any, see
-  // https://html.spec.whatwg.org/#update-the-rendering step 14.
-  NotifyResizeObservers();
-  if (MOZ_UNLIKELY(!mPresContext || !mPresContext->GetPresShell())) {
-    // A resize observer callback apparently destroyed our PresContext.
-    StopTimer();
-    return;
-  }
-
   // Update the relevancy of the content of any `content-visibility: auto`
   // elements. The specification says: "Specifically, such changes will
   // take effect between steps 13 and 14 of Update the Rendering step of
@@ -2765,6 +2755,16 @@
   // https://drafts.csswg.org/css-contain/#cv-notes
   UpdateRelevancyOfContentVisibilityAutoFrames();
 
+  // Step 14 (https://html.spec.whatwg.org/#update-the-rendering).
+  // 1) Initial proximity to the viewport determination for
+  // content-visibility:auto elements and 2) Resize observers notifications.
+  DetermineProximityToViewportAndNotifyResizeObservers();
+  if (MOZ_UNLIKELY(!mPresContext || !mPresContext->GetPresShell())) {
+    // A resize observer callback apparently destroyed our PresContext.
+    StopTimer();
+    return;
+  }
+
   UpdateIntersectionObservations(aNowTime);
 
   UpdateAnimatedImages(previousRefresh, aNowTime);